src/server.py

```diff
@@ -69,14 +69,80 @@ def response_deconstructor(response):
     headers_split = [header.split(':', 1) for header in headers]
     headers_dict = {k.lower(): v.strip() for k, v in headers_split}
     return [protocol, status, msg, headers_dict,
-            str(len(body)), str(len(first_pass))]
+            str(len(body))]
+
+
+def request_deconstructor(request):
+    """Deconstruct request into basic components.
+    Return method, protocol and validates host header"""
+    request_msg = request
+    urequest = request_msg.decode('utf8')
+    first_pass = urequest.split(CRLF+CRLF, 1)
+    head, body = first_pass
+    head_lines = head.split(CRLF)
+    method, path, protocol = head_lines[0].split()
+    headers = head_lines[1:]
+    headers_split = [header.split(':', 1) for header in headers]
+    headers_dict = {k.lower(): v.strip() for k, v in headers_split}
+    return [method, path, protocol, headers_dict, str(len(first_pass))]
+
+
+class HTTPException(Exception):
+    def __init__(self, code, reason, msg_string):
+        """Initiate an instance of HTTPException
+         with attributes: <code>, <reason>, <msg_string>."""
+        self.code = code
+        self.reason = reason
+        self.msg_string = msg_string
+
+    def response_msg(self):
+        """Generate an error message using <code>,
+         <reason>, <msg_string>. Return a byte-string"""
+        template = u'HTTP/1.1 {} {}\r\n\r\n{}'
+        return template.format(self.code, self.reason, self.msg_string)\
+            .encode('utf-8')
+
 
+def parse_req(request):
+    """Raise an appropriate error for a bad request or
+     return URI for a good request."""
+    request_decon = request_deconstructor(request)
+    if request_decon[0] != u'GET':
+        raise NotImplementedError(u'Method not allowed')
```

Comment:
Given the way your HTTPException is built, I would go ahead and raise one directly here:

```python
raise HTTPException(405, "Method Not Allowed", "405 Method Not Allowed")
```
If you do that here and in the two `elif` blocks below, it will simplify your code later.